### ISSUE TITLE
bug(runtime): agent absolute path /workspace writes to ephemeral container FS; session workspace stays empty after idle stop

### ISSUE THREAD

**Issue by Fxde42** [NeuroAIHub/BrainPilot#346]
## Summary

Agent completed a full multi-hour research task (writes + training + reports all reported success), but **nothing appears in the Files panel / session workspace** after the sandbox stops.

Root cause (verified from a real cloud session export): the agent wrote everything under the **absolute host path** `/workspace/...`, which is **not** the session workspace on the persisted volume. Idle reclaim (`docker stop`) discards the container writable layer → all artifacts vanish. The Files UI lists the managed logical root (`workspaces/<sid>/`), which stayed empty.

This is **not** #193 (panel empty while files exist on disk under `workspaces/<sid>/`), **not** #168 (sandbox-id vs session-id addressing), and **not** #293 (tool-arg truncation — all `write` calls in this session succeeded with full content).

## Environment

- Deployment: multi-user cloud (instance-per-user sandbox, single volume `{user}/.bp-root` → `/root/.bp-root`)
- Session id: `e9e92a62-ea70-4b8c-ba74-ad7352d2eb17`
- Agents involved: principal → engineer (heavy) → writer / auditor / trace
- Task: BCI Competition IV 2a EEG 4-class MI model (PyTorch)
- Provider: custom model (`preset-custom` / `custom-model-a1`)
- Observed window: 2026-07-14 → 2026-07-16 (long-running train pipeline)

### 1. Real agent cwd (correct, on volume)

First bash in the session:

```text
---cwd---
/root/.bp-root/workspaces/e9e92a62-ea70-4b8c-ba74-ad7352d2eb17
```

That is the only path the Files panel / `listSessionFiles(..., "/workspace")` should read (`session-manager.workspaceDir(sid)`).

### 2. Agent created a *second* root at absolute `/workspace`

Image has **no** `/workspace` by default. Principal explicitly:

```bash
mkdir -p /workspace/EEG_MI
touch /workspace/EEG_MI/.write_test && echo OK   # succeeded
```

That directory lives on the **container writable layer**, outside the bind mount of `BP_DATA_DIR=/root/.bp-root`.

### 3. All durable-looking work went to absolute `/workspace/...`

14 successful `write` tool calls, all with absolute paths, e.g.:

| agent | path | bytes (approx) |
|-------|------|----------------|
| engineer | `/workspace/EEG_MI/mi_agent_model.py` | 8–9k |
| engineer | `/workspace/EEG_MI/data_loader.py` | 5–9k |
| engineer | `/workspace/EEG_MI/train.py` | 7k |
| engineer | `/workspace/EEG_MI/validate_priors.py` | 16k |
| engineer | `/workspace/EEG_MI/train_segment.py` | 3k |
| writer | `/workspace/EEG_MI/results/FINAL_REPORT.md` | 7.6k |
| auditor | `/workspace/EEG_MI/results/AUDIT_REPORT.md` | 10k |

Tool results all say `Successfully wrote N bytes to /workspace/EEG_MI/...` (`is_error=false`).

Bash training also used `cd /workspace/EEG_MI` and wrote checkpoints / metrics there. End-of-run `find` listed a full tree still under `/workspace/EEG_MI/`:

- `checkpoints/S1.pt` … `S9.pt`
- `results/metrics.json`, `results/priors/*`
- sources + `logs/pipeline.log`
- downloaded `data/*.gdf` etc.

### 4. Dual semantics of `/workspace` (the actual bug)

| channel | meaning of `/workspace` | persisted? |
|---------|-------------------------|------------|
| HTTP file API / Files panel | **logical** prefix → `workspaces/<sid>/` via `resolveManagedPath` | yes (volume) |
| Pi built-in `write` / `edit` / `bash` | **real absolute path** `/workspace` on the container FS | **no** |

Personas already forbid Docker-style absolute roots (`personas.ts` design note + `personas.test.ts` asserts prompts must not contain `/workspace`). The model still used the OpenHands-style absolute root; the runtime did not rewrite or confine it.

### 5. Secondary: non-volume package installs

`pip3 install --user --break-system-packages torch ...` landed under `/root/.local` (also outside the volume). After stop, both artifacts **and** the ad-hoc env are gone. Secondary to the path issue, but same class of “container layer is not durable”.

## Expected

- Agent outputs that the model believes are “in the workspace” land under `workspaceDir(sid)` = `<BP_DATA_DIR>/workspaces/<sid>/`.
- Files panel after task completion shows those files.
- Idle stop / wake does not wipe user deliverables.

## Actual

- Deliverables lived only under absolute `/workspace` (ephemeral).
- Session workspace dir on the volume stayed empty (or without the deliverables).
- After sandbox stop, nothing recoverable via Files / volume.

## Repro sketch

1. Start sandbox (cloud or plain Docker with only `BP_DATA_DIR` mounted at `/root/.bp-root`).
2. In a session, have the agent (or manually):

   ```bash
   mkdir -p /workspace/demo
   # via write tool or echo:
   echo hello > /workspace/demo/out.txt
   pwd   # → /root/.bp-root/workspaces/<sid>
   ls    # cwd empty
   ls /workspace/demo
   ```

3. `GET /sessions/<sid>/files?path=/workspace` → empty (or no `demo/`).
4. `docker stop` the sandbox; restart/wake → `/workspace` gone; volume still empty.

(Real session export available from the cloud operator if maintainers want the full `events.jsonl`.)

## Why prompt-only is not enough

`personas.ts` already documents “no Docker mount paths (`/workspace`, `/data`, …); use cwd”. This session still created `/workspace` and wrote dozens of files there over ~2 days. Model habit (OpenHands / coding-agent conventions) beats soft prompt rules under long tool-using runs.

## Suggested directions (non-prescriptive)

Prefer a **mechanical** fix over more persona text:

1. **Path rewrite for structured tools** (`write` / `edit` / `read` / `ls` / …): map absolute prefix `/workspace` → `workspaceDir(sid)` (mirror HTTP `resolveManagedPath`). Same for documented logical `/data` if Pi tools ever see it as a bare absolute path inconsistently.
2. **Confine bash** (harder): chroot/bwrap, or pre-create a per-session bind so absolute `/workspace` cannot point at the container root layer; multi-session same container must not share one global `/workspace` (would race).
3. **Refuse / warn** on `mkdir /workspace` or writes outside `dataRoot` (+ optional `/shared` ro), surface a clear tool error: “use relative paths; cwd is the session workspace”.
4. **Optional UX**: if a successful tool result path is outside `dataRoot`, emit a system warning that the file will not survive sandbox recycle.

Avoid cloud-side “always bind host X to `/workspace`”: one container serves many sessions; a single global `/workspace` mount cannot be correct per `sid`.

## Impact

- **Severity**: high for cloud / Docker deploys — users see a finished conversation naming concrete artifacts, then an empty workspace after idle reclaim.
- **Data loss**: full pipeline outputs (models, metrics, reports) unrecoverable once the container is gone.
- Affects any model that prefers absolute `/workspace` regardless of persona wording.

## Related (distinct)

| issue | relation |
|-------|----------|
| #168 | remote Files used sandbox id — fixed; this session used correct session id semantics on the API side |
| #193 | Files empty while files **exist** under `workspaces/<sid>/` — opposite disk state |
| #293 | long `write` args truncated — here writes succeeded with full content |
| #257 / #287 | `/data` persistent library — orthogonal; this case never used the volume workspace |

## Checklist for a fix PR

- [ ] Repro test: agent (or tool harness) `write` to `/workspace/foo.txt` → file appears under `workspaces/<sid>/foo.txt`, **not** container `/workspace`
- [ ] Files list `path=/workspace` returns `foo.txt`
- [ ] After simulated stop/start with only `BP_DATA_DIR` volume, file still present
- [ ] Multi-session: two sids writing `/workspace/x` do not clobber each other
- [ ] Personas/tests remain free of advertising `/workspace` as a real Docker path (or document the rewrite explicitly)
